substrate/frame/system/src/extensions/check_nonce.rs

```diff
@@ -92,22 +104,52 @@ where
 			}
 			.into())
 		}
-		account.nonce += T::Nonce::one();
-		crate::Account::<T>::insert(who, account);
-		Ok(())
+		if !is_feeless_with_zero_nonce {
+			account.nonce += T::Nonce::one();
+			crate::Account::<T>::insert(who, account);
+		}
+		Ok((who.clone(), is_feeless_with_zero_nonce))
+	}
+
+	fn post_dispatch(
+		_pre: Option<Self::Pre>,
+		_info: &DispatchInfoOf<Self::Call>,
+		_post_info: &sp_runtime::traits::PostDispatchInfoOf<Self::Call>,
+		_len: usize,
+		_result: &sp_runtime::DispatchResult,
+	) -> Result<(), TransactionValidityError> {
+		if let Some((who, is_feeless_with_zero_nonce)) = _pre {
+			if !is_feeless_with_zero_nonce {
+				Ok(())
+			} else {
+				let mut account = crate::Account::<T>::get(who.clone());
+				if account.providers.is_zero() && account.sufficients.is_zero() {
+					return Err(InvalidTransaction::Payment.into())
+				} else {
+					account.nonce += T::Nonce::one();
+					crate::Account::<T>::insert(who, account);
```

Comment:
Would this not lead to accounts submitting feeless txns bloating the chain at no cost? There should be very clear docs explaining this trade off, and potential vulnerability here.

Reply:
> Would this not lead to accounts submitting feeless txns bloating the chain at no cost? 

Not if ED is being provided during the execution of the call. This is what is being validated in the `post_dispatch`.

> There should be very clear docs explaining this trade off, and potential vulnerability here.

Indeed. 

Open to suggestions if there is a better way to do this.

Reply:
Probably best to mention here https://paritytech.github.io/polkadot-sdk/master/frame_support/pallet_macros/attr.feeless_if.html

Reply:
> Probably best to mention here https://paritytech.github.io/polkadot-sdk/master/frame_support/pallet_macros/attr.feeless_if.html

Thanks. I meant if there are alternate approaches that could be tried :)

Reply:
Could we destroy the account after the extrinsic is processed and it's no longer needed?

Reply:
> Could we destroy the account after the extrinsic is processed and it's no longer needed?

But the feeless call might have been used for an airdrop (based on a certain check) which could have paid for the ED. This is one of the major use cases. In that case, we don't want to destroy it.

Reply:
Doesn't this change make the `SkipCheckIfFeeless` extension usable only when the underlying extrinsic does endow because of the attack vector introduced by [invalidating entire blocks on `post_dispatch`](https://github.com/paritytech/polkadot-sdk/blob/ba0f8de0c74b1308a53942c406137a748ef79925/substrate/primitives/runtime/src/traits.rs#L1576)?

Reply:
Currently, no transaction can be triggered at all when `SkipCheckIfFeeless` and `CheckNonce` are used at the same time, cause the check for nonce would cause all such transactions to be invalid. This just provides an additional mechanism to be able to at least trigger those calls where the endowment occurs during the execution. 

Runtime can devs can definitely customise their CheckNonce extension to handle this specifically, but I am not sure if there's a downside to make it a bit easier?

Reply:
> Currently, no transaction can be triggered at all when SkipCheckIfFeeless and CheckNonce are used at the same time, cause the check for nonce would cause all such transactions to be invalid.

Even transactions where the user already exists? If I have a funded account and I submit an extrinsic with a call that passes the `feeless_if` checks, does it not work? As I said here, the two problems are different and allowing nonexistent accounts to run signed transactions for free is difficult to solve and requires a robust solution because of the attack vector introduced. IMO the approach here is too big of a compromise.

Reply:
> > Could we destroy the account after the extrinsic is processed and it's no longer needed?
> 
> But the feeless call might have been used for an airdrop (based on a certain check) which could have paid for the ED. This is one of the major use cases. In that case, we don't want to destroy it.

This would bypass ED requirements, and the acc would get destroyed next ED check? :/

I think if we want to allow airdrops to new accounts, that airdrop asset needs to be sufficient otherwise we are just throwing away the point of EDs and sufficient assets..

Reply:
To me this logic looks valid. I think it even valid without `is_feeless` condition (but without it it will fail later).
But I do not like the way it complicates the general implementation used by most of the runtime. The use case is quite specific (airdrop claims). May be we provide another implementation which handles this use case.

Reply:
> To me this logic looks valid. I think it even valid without `is_feeless` condition (but without it it will fail later). But I do not like the way it complicates the general implementation used by most of the runtime. The use case is quite specific (airdrop claims). May be we provide another implementation which handles this use case.

Yes, I was thinking to implement a wrapper similar to the `SkipCheckIfFeeless` extension itself such that the wrapped `CheckNonce` will only be called if the feeless conditions are met and the nonce is 0. This would remove the coupling here, while providing an extensible mechanism.